[Refactor] Team-04 리뷰 내용 반영 및 중복 오류 수정

longrunBiin opened the pull request:
### 1. 작업 내용
- EOF, 띄어쓰기 수정
- 리팩토링 : IssueService에서 쿼리 작성하던 부분을 Dao로 분리
- null체크만을 위한 Optional을 제거
- getIssue() 데이터 중복 출력 문제 해결

### 2. 이번 작업 달라진 점
- createIssue()에서 마일스톤을 null체크 하던 부분을 제거하였습니다.
- S3FileService에서 반환값을 Optional로 바꿔서 uploadUrl이 없는 경우를 명확히 했습니다.
- IssueService.getIssue()에서 데이터가 중복되던 문제를 List대신 Set을 사용하고 Dto에서 equals(), hashcode()를 구현하여 중복을 제거하였습니다.

### 3. 고민사항
- IssueService의 쿼리 부분을 Dao로 분리를 한 이유는 createIssue()와 의존하는 레벨이 달랐기 때문입니다. createIssue()는 Issue를 저장하기 위해 CrudRepository를 상속받은 IssueRepository를 의존하고 있었고 getIssue()는 JdbcTemplate을 의존하고 있었기 때문에 하나의 서비스에서 의존하고 있는 객체의 계층이 다르다고 생각이 들었습니다. 
 또한 테스트에서도 JdbcTemplate을 목객체로 사용해서 쿼리를 작성하고 given()..willReturn()에 넣다보니 불편합이 컸습니다.

- getIssue()를 어떻게 리팩토링 할지를 고민 중입니다. 단순히 메소드를 나누기 보다는 각 기능에 맞는 메소드와 클래스로 분리해보려고 고민중입니다.

- 로그를 사용하는 방법에 고민이 있습니다. 언제 log.info를 사용하고 언제 log.debug를 사용할지 로그를 어떤 경우에 찍어야 할지 등 정해진 답이 없다보니 디버깅을 하는게 아니면 잘 안찍게 되는 것 같습니다.

longrunBiin committed:
chore : EOF, 띄어쓰기 수정

longrunBiin committed:
refactor : 쿼리 작성부분 Dao로 분리

longrunBiin committed:
refactor : IssueCountService 병합
- IssueCountService -> IssueService.getIssueCounts로 이동
- IssueCountService 삭제
- 쿼리 부분 IssueService -> IssueDao 이동으로 인한 테스트 수정

longrunBiin committed:
refactor : Service 뭐리 부분 Dao로 이동
- LabelService, MilestoneService, UserService의 쿼리 부분 각각의 Dao로 이동
- Dto 생성 Builder로 통일

longrunBiin committed:
fix : 테스트 의존 객체 변경
- JdbcTemplate -> Dao객체로 변경

longrunBiin committed:
refactor : MilestoneDto.MilestoneInfo 필드 변경
- name -> title로 변경

longrunBiin committed:
fix : IssueService.getIssues 수정
- 레이블, 담당자 중복 출력 막기 위해 assigneeMap, labelMap Map<Long, List> -> Map<Long, Set>으러 변경
- 동등성 보장을 위해 LabelDto.LabelInfo, UserDto.UserInfo equals(), hashcode() 오버라이딩

longrunBiin committed:
chore : gitignore 추가
- .idea 파일 추가
- eof 추가

longrunBiin committed:
refactor : IssueService.createIssue milestone null 체크 제거
- 의미없는 Optional, null 체크 제거
- DTO 필수 필드에 @NonNull 추가

longrunBiin committed:
refactor : S3FileService.uploadFile Optional을 반환하도록 수정
- null 체크만을 하는 것이 아닌 Optional을 반환하도록 수정

moongyuxx committed:
Merge pull request #25 from codesquad-masters2025-team04/refactor/issueService

[Refactor] IssueService 리팩토링 및 리뷰 반영

nag0124 commented on backend/src/main/java/codesquad/team4/issuetracker/issue/dto/IssueRequestDto.java:
```diff
@@ -4,15 +4,19 @@
     @Getter
     @Builder
     public static class CreateIssueDto{
+        @NonNull
```
(조언)

이 로직은 requstBody를 검증하는 로직이네요.
requestBody를 검증할 떄는 spring validation의 `@NotNull`을 사용하셔도 좋을 것 같습니다. 한 번 학습해보세요.

longrunBiin replied:
`@NotNull`을 사용해서 컨트롤러단에서 request를 검증하는 게 더 좋은 방법인 것 같습니다!

nag0124 commented on backend/src/main/java/codesquad/team4/issuetracker/issue/IssueService.java:
```diff
@@ -121,17 +98,13 @@ public IssueResponseDto.IssueListDto getIssues(boolean isOpen, int page, int siz
 
-            builder.assignees(assigneeMap.getOrDefault(issueId, List.of()));
-            builder.labels(labelMap.getOrDefault(issueId, List.of()));
+            builder.assignees(List.copyOf(assigneeMap.getOrDefault(issueId, Set.of())));
```
(질문)

여기서 List.copyOf()를 사용하여 Set을 반환하지 않고 List로 반환하시는 특별한 이유가 있나요?
없다면 응답으로 Set을 줘도 괜찮을 것 같습니다.

longrunBiin replied:
DTO에 List로 선언을 해놨기 때문에 List.copyOf()를 사용했었는데 지금 생각하니 그냥 Set을 반환해버리는게 맞다고 생각됩니다

nag0124 reviewed (changes requested):
이번 주 고생 많으셨습니다.
코드 대부분이 리뷰 반영인 거 같은데 제가 리뷰를 과도하게 한 건 아닌지 걱정이군요...
그리고 제가 첫 리뷰 떄 놓친 내용들이 있어서 몇가지 좀 적겠습니다.

### 1. S3에 이미지 업로드시 다른 예외 처리 (조언)
지금 로직상에서는 `IOException`만 처리를 하고 있는데, `S3Client.putObject()` 메서드를 호출하면 더 많은 예외가 발생할 수 있습니다.
<img width="1143" alt="image" src="https://github.com/user-attachments/assets/203f9820-f3f6-4dc0-82c6-870526d3c3de" />
이 예외 케이스를 다 구분하는 건 시간이 오래 걸릴 것 같고, `IOException`을 처리하는 걸 `Exception`을 처리하는 걸로 변경하여 일괄적으로 처리하고 추후 시간이 나면 세분화 하는게 좋을 것 같습니다.

### 2. 이미지 업로드 실패시 예외 응답 리턴 (질문)
이슈를 생성할 때 파일 업로드를 실패한다면 이슈 자체를 만들지 않고 실패 응답을 내보내고 있어요.
이 경우 이슈 생성 기능이 파일 업로드라는 외부 환경에 의존적이게 됩니다.
만약 AWS에 잠시 동안 장애가 발생한다면 그 시간 동안의 이슈 생성 자체가 불가합니다. 
이슈 생성시 파일 업로드가 필수인지 선택인지 고민해보고 알려주세요.

### 3. 로컬 서버 설정은 어떻게 생각하시는가요? (조언)
먼저 로컬 서버를 띄울 수 있는지 없는지 궁금합니다.
제가 확인했을 때는 로컬 서버를 띄울 수 있는 정보가 없는 걸로 보이는데, 팀적으로 공유하고 계신건지 궁금합니다.
로컬 서버를 띄우는게 좋을 수도 있는 게, 실제 배포를 하기 전에 로컬에서 서버를 띄우고 여러 요청을 보내면서 내가 의도한 대로 코드가 동작하는지 테스팅을 할 수 있다는 장점이있습니다.
배포 서버까지 가서 테스팅을 하다보면 에러가 날 떄마다 코드를 고치고 배포를 하고 고쳐졌는지 확인하는 작업이 필요하거든요.
테스트 DB를 H2db를 사용하신다면, 비슷한 방법을 써서 로컬 서버도 띄우는 걸 추천합니다.

### 4. 로그는 어떻게 띄워야 할까요? (조언)
제 생각에 로그는 서버가 개발자에게 보여주는 메세지입니다.
종류가 trace, debug, info, warn, error가 있는데, 지금은 info와 error만 쓰셔도 무방할 거 같아요.
trace를 찍으면 로그가 너무 많아지고, debug는 로컬 서버에서 디버거로 대체가 가능하다고 생각하고 warn는 잠재적인 위험이 있을 때 찍는 데 지금 쓰기엔 애매한 것 같아요.
info와 error만 설명해드리면 
- info: flow를 잘 타고 있다는 걸 보여주는 용도(API 호출됨, 이미지 업로드 완료됨 등)
- error: 예외 상황(이미지 업로드 실패)

로그를 언제 찍어야 하냐에 대한 답은 저도 잘 모르겠지만, 이걸 고려하셔서 찍으시면 좋을 것 같습니다.
"서버가 어떤 상황에서 정보를 알려주는게 좋을까?"

longrunBiin committed:
fix : 매핑 테이블 updatedAt 제거, 외래키 변경
- IssueLabel, IssueAssignee updatedAt 제거
- Issue.commentId 제거 -> comment.issueId 추가

longrunBiin committed:
refactor : IssueDao.findIssuesByOpenStatus issueLabel, issueAssignee left join -> inner join으로 수정

longrunBiin committed:
refactor : IssueService.createIssue null체크를 위한 Optional 제거
- if문 null체크로 수정

longrunBiin committed:
refactor : ddl 수정

longrunBiin committed:
chore : S3 접근 방식 IAM ROLE로 변경

longrunBiin committed:
refactor : IssueResponseDto.IssueInfo labels, assignees List -> Set 변경

longrunBiin committed:
refactor : IssueRequestDto.CreateIssueDto @NonNull -> NotNull 수정
- IssueController.createIssue @Valid로 DTO null 검사
- validation 의존성 추가

longrunBiin committed:
refactor : S3FileService.uploadFile IOException -> Exception 으로 수정
- 추후 디테일한 에러 핸들링 수정

longrunBiin commented:
배포 전 테스트를 위한 로컬서버는 각자의 로컬 환경에서 프로그램을 실행 시키는 방법으로 진행을 했는데 안그래도 좀 불편함을 느끼던 중이었습니다 
생각했던 방법은 테스트 서버를 ec2로 띄워서 로컬에서 접근하고 테스트를 하게끔 할지 고민을 하고 있습니다
하지만 아무래도 서버 비용에 대한 문제와 초기 세팅의 번거로움 때문에 로컬 서버를 사용하는게 좋을 것 같다는 생각도 들어서 고민 중입니다

longrunBiin commented:
이미지 업로드 실패시 예외 응답을 리턴 하는 경우는 
 이미지가 현재 이슈에 필수가 아닌데 이미지 업로드를 실패했다고 이슈 생성이 되지 않는 것은 잘못된 로직인 것 같습니다
에러 발생시에도  이슈를 정상적으로 생성하고 에러메세지를 추가하는 식으로 바꾸려고 합니다

nag0124 reviewed (commented):
고생하셨습니다!
